File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        vector<bool> visited(arr.size(), false);
        return dfs(arr, visited, start);
    }
private:
    bool dfs(vector<int>& arr, vector<bool>& visited, int start)
    {
        if(start < 0 || start >= arr.size() || visited[start])
        {
            return false;
        }
        visited[start] = true;
        if(arr[start] == 0)
        {
            return true;
        }
        return (dfs(arr, visited, start + arr[start]) || dfs(arr, visited, start - arr[start]));
    }
};

File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        visited = [False for _ in range(len(arr))]
        def dfs(start: int) -> bool:
            if start < 0 or start >= len(arr) or visited[start]:
                return False
            visited[start] = True
            if arr[start] == 0:
                return True
            return dfs(start + arr[start]) or dfs(start - arr[start])    
        return dfs(start)